src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java

```diff
@@ -101,6 +101,7 @@ public class InnerAssignmentCheck
             TokenTypes.RESOURCE_SPECIFICATION,
         },
         {TokenTypes.EXPR, TokenTypes.LAMBDA},
+        {TokenTypes.EXPR, TokenTypes.SWITCH_RULE, TokenTypes.LITERAL_SWITCH, TokenTypes.SLIST},
```

Comment:
Why we need TokenTypes.SLIST ? And TokenTypes.LITERAL_SWITCH

Reply:
To make sure that we are in normal Switch context not Switch Expression. if it is  `{TokenTypes.EXPR, TokenTypes.SWITCH_RULE}` only then we will have no violation if the assignment is in switch expression
ex 
```java
    public void test2(int mode) {
        int x = 0;
        x = switch (mode) {
            case 2 -> {
                yield x = 2; 
            }
            case 1 -> x = 1;  // this will be ok
            default -> throw new UnsupportedOperationException();
        };
    }
``` 

Switch Expression AST:
```
        |--EXPR -> EXPR [8:10]
        |       |   `--ASSIGN -> = [8:10]   <<---
        |       |       |--IDENT -> x [8:8]
        |       |       `--LITERAL_SWITCH -> switch [8:12]   <<---
        |       |           |--LPAREN -> ( [8:19]
        |       |           |--EXPR -> EXPR [8:20]
        |       |           |   `--IDENT -> mode [8:20]
        |       |           |--RPAREN -> ) [8:24]
        |       |           |--LCURLY -> { [8:26]
        |       |           |--SWITCH_RULE -> SWITCH_RULE [9:12]   <<---
        |       |           |   |--LITERAL_CASE -> case [9:12]
        |       |           |   |   `--EXPR -> EXPR [9:17]
        |       |           |   |       `--NUM_INT -> 1 [9:17]
        |       |           |   |--LAMBDA -> -> [9:19]
        |       |           |   |--EXPR -> EXPR [9:24]         <<---
        |       |           |   |   `--ASSIGN -> = [9:24]
        |       |           |   |       |--IDENT -> x [9:22]
        |       |           |   |       `--NUM_INT -> 1 [9:26]
        |       |           |   `--SEMI -> ; [9:27]
        |       |           `--RCURLY -> } [10:8]
```
switch rule AST

```
        `--SLIST -> { [6:33]          <<---
        |       |--LITERAL_SWITCH -> switch [7:8]   <<---
        |       |   |--LPAREN -> ( [7:15]
        |       |   |--EXPR -> EXPR [7:16]
        |       |   |   `--IDENT -> mode [7:16]
        |       |   |--RPAREN -> ) [7:20]
        |       |   |--LCURLY -> { [7:22]
        |       |   |--SWITCH_RULE -> SWITCH_RULE [8:12]          <<---
        |       |   |   |--LITERAL_CASE -> case [8:12]
        |       |   |   |   `--EXPR -> EXPR [8:17]
        |       |   |   |       `--NUM_INT -> 1 [8:17]
        |       |   |   |--LAMBDA -> -> [8:19]
        |       |   |   |--EXPR -> EXPR [8:24]       <<---
        |       |   |   |   `--ASSIGN -> = [8:24]
        |       |   |   |       |--IDENT -> x [8:22]
        |       |   |   |       `--NUM_INT -> 1 [8:26]
        |       |   |   `--SEMI -> ; [8:27]
        |       |   `--RCURLY -> } [9:8]

```

Reply:
@romani ping